File: Day-06-Word-Pattern/Solution.java
class Solution {
    public boolean wordPattern(String pattern, String str) {
        Map<String, String> hashMap = new HashMap<>();

        String[] strArr = str.split(" ");
        if(pattern.length() != strArr.length)
            return false;
        
        for(int i = 0; i < pattern.length(); i++) {
            String s = String.valueOf(pattern.charAt(i));
            if( hashMap.containsKey(s) && !hashMap.get(s).equals(strArr[i]) ) {
                    return false;
            } else {
                hashMap.put(s, strArr[i]);
            }
            
            if(i > 0) {
                String st = String.valueOf(pattern.charAt(i - 1));
                if( !st.equals(s) && hashMap.get(s).equals(hashMap.get(st)) )
                    return false;
            }
        }
        
        return true;
    }
}

File: Day-06-Word-Pattern/Solution.cpp
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        istringstream iss(str);
        vector<string> strArr{istream_iterator<string>{iss},
                             istream_iterator<string>{}};
        
        if(pattern.length() != strArr.size())
            return false;
        
        unordered_map<char, string> mapOne;
        unordered_map<string, char> mapTwo;
        
        for(int i = 0; i != pattern.length(); ++i) {
            unordered_map<char, string>::const_iterator one = mapOne.find(pattern[i]);
            unordered_map<string, char>::const_iterator two = mapTwo.find(strArr[i]);
            
            if( one == mapOne.end() && two == mapTwo.end() ) {
                mapOne[pattern[i]] = strArr[i];
                mapTwo[strArr[i]] = pattern[i];
                continue;
            } 
            
            if( mapOne[pattern[i]] != strArr[i] )
                return false;
            
        }
        return true;
    }
};
